sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -1163,4 +1163,40 @@ object DateTimeUtils {
     val localStartTs = getLocalDateTime(startMicros, zoneId)
     ChronoUnit.MICROS.between(localStartTs, localEndTs)
   }
+
+  /**
+   * Adds the specified number of units to a timestamp.
+   *
+   * @param unit A keyword that specifies the interval units to add to the input timestamp.
+   * @param quantity The amount of `unit`s to add. It can be positive or negative.
+   * @param micros The input timestamp value, expressed in microseconds since 1970-01-01 00:00:00Z.
+   * @param zoneId The time zone ID at which the operation is performed.
+   * @return A timestamp value, expressed in microseconds since 1970-01-01 00:00:00Z.
+   */
+  def timestampAdd(unit: String, quantity: Int, micros: Long, zoneId: ZoneId): Long = {
+    unit.toUpperCase(Locale.ROOT) match {
+      case "MICROSECOND" =>
+        timestampAddDayTime(micros, quantity, zoneId)
+      case "MILLISECOND" =>
+        timestampAddDayTime(micros, quantity * MICROS_PER_MILLIS, zoneId)
+      case "SECOND" =>
+        timestampAddDayTime(micros, quantity * MICROS_PER_SECOND, zoneId)
+      case "MINUTE" =>
+        timestampAddDayTime(micros, quantity * MICROS_PER_MINUTE, zoneId)
+      case "HOUR" =>
+        timestampAddDayTime(micros, quantity * MICROS_PER_HOUR, zoneId)
+      case "DAY" | "DAYOFYEAR" =>
+        timestampAddDayTime(micros, quantity * MICROS_PER_DAY, zoneId)
+      case "WEEK" =>
+        timestampAddDayTime(micros, quantity * MICROS_PER_DAY * DAYS_PER_WEEK, zoneId)
+      case "MONTH" =>
+        timestampAddMonths(micros, quantity, zoneId)
+      case "QUARTER" =>
+        timestampAddMonths(micros, quantity * 3, zoneId)
+      case "YEAR" =>
+        timestampAddMonths(micros, quantity * MONTHS_PER_YEAR, zoneId)
+      case _ =>
+        throw QueryExecutionErrors.invalidUnitInTimestampAdd(unit)
```

Comment:
Shall we check the unit names in the parser? To fail earlier.

Reply:
Since we don't the optimization for foldable unit names. It doesn't make sense.

Reply:
We can but I just wonder what kind of problem would the earlier check solve? Parser and compiler do a lot of work for now, adding more unnecessary things should be motivated somehow, from my point of view.

Reply:
Failing earlier is a pretty strong reason, right? It's a waste of resource if we submit a Spark job which fails with wrong unit name.

Reply:
>  It's a waste of resource if we submit a Spark job which fails with wrong unit name.

Not sure. Can you imagine a cluster of 1000 executors waiting for the driver that is still processing a query because we eagerly want to check everything even when user's queries and data don't have any issues. This is real waste of user's resources.

Reply:
Also I would like to add, what are you taking is about a mistake in a query actually, like:
```sql
SELECT timestampadd(YEER, 1, timestampColumn);
```
Such kind of mistakes are not permanent, and usually users fix them during the debug stage. There are no so much reasons to double check such mistakes at parsing and in runtime (we must do that since `unit` can be non-foldable).

Reply:
But this violates the convention: we try to fail as early as possible in all other places. I don't understand why we suddenly want to stop doing it from this PR. It's just a few CPU cycles in this case and I don't see how it's a performance overhead for the driver.

Reply:
> I don't understand why we suddenly want to stop doing it from this PR.

1. The `unit` param can be non-foldable. I made it generic intentionally. If you wonder why, I will answer to that separately.
2. As `unit` can be non-foldable, we need the runtime check.
3. If we add checks in parser, we will do checks twice at parsing and at execution... which is not necessary because
4. We can handle foldable `unit` in codegen as an optimization where we (of course) have to check `unit` values at the optimization phase.

As summary, taking into account that we will optimize foldable `unit` in codegen in the near future where we validate correctness of `unit`, there is no need to do that in parser as you proposed.

> Example: EXTRACT, TO_BINARY, TO_NUMBER

The expressions require one of their param (format, field and etc) to be **always** foldable. In the case, of `TIMESTAMPADD()` is unnecessary restriction, I believe. I have faced to the situation a few times in my life when some code was deployed in the production after testing, and need to increase precision of intervals. Let's say we had:
```sql
select timestampadd(SECOND, tbl.quantity, tbl.ts1) 
```
, and we wants to bump precision of `tbl.quantity` to milliseconds. Since `quantity` is a column in a table, we can just multiply it by 1000 during a maintenance time but we should do with `SECOND`?  We have to re-deploy to code, including pass whole release cycle, only because a Spark dev forced us to hard-code the `SECOND` in our code, for some unclear reasons.

Reply:
I'm totally OK with your decision if the unit parameter can be unfoldable, but it seems not the case? We even added a special parse rule for this function so that the unit parameter is an identifier.

Reply:
Technically speaking the first argument, unit, should be a datetime interval type as in what's used with EXTRACT. It isn't meant to be a string if that makes things any simpler.

Reply:
> the first argument, unit, should be a datetime **interval type**

I didn't get your point. How it could be the interval type?

> ... as in what's used with EXTRACT

Just wonder why do you linked `TIMESTAMPADD` to `EXTRACT` but not to `TIMESTAMPDIFF`, for example.  Anyway technically specking the type of the first argument is the same - string type.

> ... makes things any simpler.

This PR achieve this goal, I believe. It makes the migration process to Spark SQL simpler, and gives additional benefits of using Spark SQL in the real production (see my comment above).